dg.go

```diff
@@ -1,315 +1,464 @@
+func (n *node[NodeType]) DisconnectInbound(fromNodeID string) error {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	if n.deleted {
+		return &ErrNodeDeleted{n.id}
+	}
+	if _, ok := n.dg.nodes[fromNodeID]; !ok {
+		return &ErrNodeNotFound{fromNodeID}
+	}
+	if _, ok := n.dg.connectionsToNode[n.id][fromNodeID]; !ok {
+		return &ErrConnectionDoesNotExist{n.id, fromNodeID}
+	}
+	delete(n.dg.connectionsToNode[n.id], fromNodeID)
+	delete(n.dg.connectionsFromNode[fromNodeID], n.id)
+	return nil
 }
 
-func (d *node[NodeType]) Item() NodeType {
-    return d.item
+func (n *node[NodeType]) DisconnectOutbound(toNodeID string) error {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	if n.deleted {
+		return &ErrNodeDeleted{n.id}
+	}
+	if _, ok := n.dg.nodes[toNodeID]; !ok {
+		return &ErrNodeNotFound{toNodeID}
+	}
+	if _, ok := n.dg.connectionsFromNode[n.id][toNodeID]; !ok {
+		return &ErrConnectionDoesNotExist{n.id, toNodeID}
+	}
+	delete(n.dg.connectionsFromNode[n.id], toNodeID)
+	delete(n.dg.connectionsToNode[toNodeID], n.id)
+	return nil
 }
 
-func (d *node[NodeType]) Connect(nodeID string) error {
-    d.dg.lock.Lock()
-    defer d.dg.lock.Unlock()
-    if d.deleted {
-        return &ErrNodeDeleted{
-            d.id,
-        }
-    }
-    if nodeID == d.id {
-        return &ErrCannotConnectToSelf{
-            nodeID,
-        }
-    }
-    if _, ok := d.dg.nodes[nodeID]; !ok {
-        return &ErrNodeNotFound{
-            nodeID,
-        }
-    }
-    if _, ok := d.dg.connectionsFromNode[d.id][nodeID]; ok {
-        return &ErrConnectionAlreadyExists{
-            d.id,
-            nodeID,
-        }
-    }
-    d.dg.connectionsFromNode[d.id][nodeID] = struct{}{}
-    d.dg.connectionsToNode[nodeID][d.id] = struct{}{}
-    return nil
+func (n *node[NodeType]) Remove() error {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	if n.deleted {
+		return &ErrNodeDeleted{n.id}
+	}
+	for toNodeID := range n.dg.connectionsFromNode[n.id] {
+		delete(n.dg.connectionsToNode[toNodeID], n.id)
+	}
+	delete(n.dg.connectionsFromNode, n.id)
+	for fromNodeID := range n.dg.connectionsToNode[n.id] {
+		delete(n.dg.connectionsFromNode[fromNodeID], n.id)
+	}
+	delete(n.dg.connectionsToNode, n.id)
+	delete(n.dg.nodes, n.id)
+	n.deleted = true
+	return nil
 }
 
-func (d *node[NodeType]) DisconnectInbound(fromNodeID string) error {
-    d.dg.lock.Lock()
-    defer d.dg.lock.Unlock()
-    if d.deleted {
-        return &ErrNodeDeleted{
-            d.id,
-        }
-    }
-    if _, ok := d.dg.nodes[fromNodeID]; !ok {
-        return &ErrNodeNotFound{
-            fromNodeID,
-        }
-    }
-    if _, ok := d.dg.connectionsToNode[d.id][fromNodeID]; !ok {
-        return &ErrConnectionDoesNotExist{
-            d.id,
-            fromNodeID,
-        }
-    }
-    delete(d.dg.connectionsToNode[d.id], fromNodeID)
-    delete(d.dg.connectionsFromNode[fromNodeID], d.id)
-    return nil
+func (n *node[NodeType]) ListInboundConnections() (map[string]Node[NodeType], error) {
```

Comment:
It's slightly weird that a function named "List" would return a _map_.

Also, `n.dg.nodes[fromNodeID]` is (I think...) a pointer, but we're returning objects.  Apparently, that's not a problem, but it seems...curious.

The same comments apply to `ListOutboundConnections()`.

Reply:
It's an oddity, but it's the public API that cannot change. This part only has formatting changes compared to main.